store/badger.go

```diff
@@ -0,0 +1,177 @@
+package store
+
+import (
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/dgraph-io/badger/v3"
+	"github.com/sirupsen/logrus"
+)
+
+type BadgerStore struct {
+	db       *badger.DB
+	gcExitCh chan struct{}
+	wg       sync.WaitGroup
+}
+
+type badgerLogger struct{}
+
+func (*badgerLogger) Errorf(f string, a ...interface{}) {
+	logrus.Errorf(f, a...)
+}
+
+func (*badgerLogger) Warningf(f string, a ...interface{}) {
+	logrus.Warningf(f, a...)
+}
+
+func (*badgerLogger) Infof(f string, a ...interface{}) {
+	logrus.Infof(f, a...)
+}
+
+func (*badgerLogger) Debugf(f string, a ...interface{}) {
+	logrus.Debugf(f, a...)
+}
+
+func logrusLevelToBadgerLevel(options badger.Options) badger.Options {
+	switch logrus.GetLevel() {
+	case logrus.InfoLevel:
+		return options.WithLoggingLevel(badger.INFO)
+	case logrus.TraceLevel:
+		fallthrough
+	case logrus.DebugLevel:
+		return options.WithLoggingLevel(badger.DEBUG)
+	case logrus.WarnLevel:
+		return options.WithLoggingLevel(badger.WARNING)
+	case logrus.FatalLevel:
+		fallthrough
+	case logrus.PanicLevel:
+		fallthrough
+	case logrus.ErrorLevel:
+		return options.WithLoggingLevel(badger.ERROR)
+	default:
+		return options.WithLoggingLevel(badger.ERROR)
+	}
+}
+
+func NewBadgerStore(path string, userID string, encryptionPassphrase []byte) (*BadgerStore, error) {
+	directory := filepath.Join(path, userID)
+	db, err := badger.Open(logrusLevelToBadgerLevel(badger.DefaultOptions(directory)).
+		WithLogger(&badgerLogger{}).
+		WithEncryptionKey(encryptionPassphrase).
+		WithIndexCacheSize(128 * 1024 * 1024))
+
+	if err != nil {
+		return nil, nil
+	}
+
+	store := &BadgerStore{db: db}
```

Comment:
can we create `gcExitCh` here?

and in `func (b *BadgerStore) Close() error`, close the channel. Then the GC goroutine will exit (reads on closed channels always succeed), and you don't need to create/close the channel in the `startGCCollector` function

Reply:
Good call. Fixed.